base/loading.jl

```diff
@@ -3333,6 +3341,65 @@ function resolve_depot(inc::AbstractString)
     return :no_depot_found
 end
 
+"""
+    Base.RelocPath(path::AbstractString)
+
+A type to represent a relocatable path.
+
+Requires `path` to be located within one of `DEPOT_PATH` upon construction.
+
+An error is thrown if relocation fails.
+
+# Example
+```jldoctest
+julia> path1 = joinpath(mktempdir(), "foo"); touch(path1); # set up a file called foo
+
+julia> pushfirst!(DEPOT_PATH, dirname(path1));
+
+julia> relocpath = Base.RelocPath(path1);
+
+julia> String(relocpath) == path1
+true
+
+julia> path2 = joinpath(mktempdir(), "foo"); touch(path2); # set up another foo
+
+julia> pushfirst!(DEPOT_PATH, dirname(path2));
+
+julia> String(relocpath) == path2
+true
+
+julia> path1 != path2
+true
+```
+"""
+struct RelocPath
+    subpath::String
```

Comment:
Thoughts about using `PerProcess` to make it so we only compute the result once:
```suggestion
struct RelocPath
    getpath::PerProcess{String}
end
RelocPath(path::AbstractString) = let
    depot, _ = replace_depot_path_impl(path)
    if isnothing(depot)
        error("Failed to locate $(path) in any of DEPOT_PATH.")
    end
    subpath = String(replace(path, depot => ""; count=1))
    return RelocPath(PerProcess{String}() do
        for d in DEPOT_PATH
            if isdirpath(d)
                d = dirname(d)
            end
            local path = string(d, subpath)
            if ispath(path)
                return path
            end
        end
        error("Failed to relocate @depot$(subpath) in any of DEPOT_PATH.")
    end)
end
String(r::RelocPath) = r.getpath()
```

Reply:
I don't love this. What if someone has modified `Base.DEPOT_PATH` during their Julia session?